modules/client/src/main/java/org/apache/ignite/internal/client/table/AbstractClientView.java

```diff
@@ -135,21 +123,28 @@ public Cursor<T> query(@Nullable Transaction tx, @Nullable Criteria criteria, @N
     public CompletableFuture<AsyncCursor<T>> queryAsync(
             @Nullable Transaction tx,
             @Nullable Criteria criteria,
-            CriteriaQueryOptions opts
+            @Nullable CriteriaQueryOptions opts
     ) {
+        CriteriaQueryOptions opts0 = opts == null ? CriteriaQueryOptions.DEFAULT : opts;
+
         return tbl.getLatestSchema()
                 .thenCompose((schema) -> {
                     SqlSerializer ser = createSqlSerializer(tbl.name(), schema.columns(), criteria);
+
+                    Statement statement = new ClientStatementBuilder().query(ser.toString()).pageSize(opts0.pageSize()).build();
                     Session session = new ClientSessionBuilder(tbl.channel()).build();
 
-                    return session.executeAsync(tx, ser.toString(), ser.getArguments())
-                            .thenApply(resultSet -> {
-                                ResultSetMetadata meta = resultSet.metadata();
+                    return doWithCallbackOnFailure(
```

Comment:
Why can't you just write .exceptionally(ex -> {
            session.close();
            ExceptionUtils.sneakyThrow(ex); // Need to think about convert to right public exception
            return null;
            );'
            ?

Reply:
An exception can be thrown before the future is created, but it doesn't seem to be this case. Replaced with `exceptionally`. Convert to right public exception will be done in https://issues.apache.org/jira/browse/IGNITE-21146